src/Dapr.AspNetCore/CloudEventsMiddleware.cs

```diff
@@ -70,32 +71,33 @@ private async Task ProcessBodyAsync(HttpContext httpContext, string charSet)
             string originalContentType;
             string contentType;
 
-            // Data is optional.
-            if (json.TryGetProperty("data", out var data))
+            // Check whether to use data or data_base64 as per https://github.com/cloudevents/spec/blob/v1.0.1/json-format.md#31-handling-of-data
+            var isDataSet = json.TryGetProperty("data", out var data);
+            var isBinaryDataSet = json.TryGetProperty("data_base64", out var binaryData);
+
+            if (isDataSet && isBinaryDataSet)
+            {
+                await httpContext.Response.WriteAsync("Both, data and data_base64 fields set in the Cloudevents envelope. Invalid request");
```

Comment:
I would suggest not doing this. Putting some content inside the request often raises more questions than it answers - like how would a machine understand this? What's the right format, etc. It's better to just avoid having an opinion on this.

If you want to make this visible to the **developer** of the application, then logging is the right way to do that. You could do that by injecting an `ILogger<CloudEventsMiddleware>` in the constructor.